src/plugins/directoryvalue/directoryvalue_delegate.cpp

```diff
@@ -225,18 +225,26 @@ KeySetPair splitEmptyArrayParents (kdb::KeySet const & arrayParents)
 	for (auto arrayParent : arrayParents)
 	{
 		kdb::Key parent = arrayParent.dup ();
-
-		parent.rewindMeta ();
-		size_t metaSize = 0;
+		// Works also without rewindMeta(), but not without nextMeta() in the inner loop...
 		bool isEmpty = parent.getBinarySize () == 0;
-		while (isEmpty && parent.nextMeta ())
+
+		ckdb::KeySet * metaKeys = ckdb::keyMeta (parent.getKey ());
+		for (ssize_t metaSize = 0; isEmpty && metaSize < ckdb::ksGetSize (metaKeys); metaSize++)
 		{
-			if (metaSize > 2 || parent.currentMeta ().getName () != "binary" || parent.currentMeta ().getName () != "array")
+			/* TODO: Seems to only work with keyNextMeta, usage of external iterator produces additional '__dirdata:' entries
+			 * in files when using the yajl plugin with arrays! */
+			const ckdb::Key * curMeta = ckdb::keyNextMeta (parent.getKey ());
+
+			/* produces additional __dirdata entry in json file */
+			// const ckdb::Key * curMeta = ckdb::ksAtCursor (metaKeys, metaSize);
+
+			if (metaSize > 2 || (curMeta && (strcmp (ckdb::keyName (curMeta), "meta:/binary") ||
+							 strcmp (ckdb::keyName (curMeta), "meta:/array"))))
 			{
 				isEmpty = false;
 			}
-			metaSize++;
 		}
+
 		(isEmpty ? emptyParents : nonEmptyParents).append (arrayParent);
 	}
```

Comment:
The original code is very dense, but if I understood it correctly, the old `for` loop is equivalent to this more readable variant:

```cpp
	for (auto arrayParent : arrayParents)
	{
		if (arrayParent.getBinarySize() != 0)
		{
			nonEmptyParents.append(arrayParent);
			continue;
		}

		// dup needed to preserve internal cursor because we use rewindMeta/nextMeta
		kdb::Key parent = arrayParent.dup ();

		parent.rewindMeta ();
		size_t metaSize = 0;
		while (metaSize < 2 && parent.nextMeta () != NULL)
		{
			if (parent.currentMeta ().getName () != "binary" || parent.currentMeta ().getName () != "array")
			{
				// mark arrayParent as non-empty
				metaSize = 3;
				break;
			}
			metaSize++;
		}

		// arrayParent is "empty" if it has no value and at most the metakeys array and/or binary
		if (metaSize < 2)
		{
			emptyParents.append(arrayParent);
		}
		else
		{
			nonEmptyParents.append(arrayParent);
		}
	}
```

Under this assumption, it should be possible to rewrite this without the inner loop entirely:

```cpp
	for (auto arrayParent : arrayParents)
	{
		if (arrayParent.getBinarySize () != 0 || !arrayParent.hasMeta("array"))
		{
			// has value, or is not actually array parent -> non-empty
			nonEmptyParents.append (arrayParent);
			continue;
		}

		ckdb::KeySet * metaKeys = ckdb::keyMeta(arrayParent.getKey());
		if (arrayParent.hasMeta("binary") && ckdb::ksGetSize(metaKeys) > 2)
		{
			// has meta:/binary and at least 2 other metakeys (incl. meta:/array) -> non-empty
			nonEmptyParents.append (arrayParent);
			continue;
		}

		if (ckdb::ksGetSize(metaKeys) > 1)
		{
			// no meta:/binary, but at least 2 other metakeys (incl. meta:/array) -> non-empty
			nonEmptyParents.append (arrayParent);
			continue;
		}

		// only one metakey, which must be meta:/array -> empty
		emptyParents.append(arrayParent);
	}
```

If this doesn't fix the problem, either I misunderstood what the function should do, or the problem is elsewhere entirely

Reply:
Thanks, that is clearly more easy to understand!
Unfortunately, the problem when not using `keyNextMeta()` is also present in the new variant.

I just determined the number of metakeys with two different approaches:
```cpp
// use ksGetSize()
ckdb::KeySet * metaKeys = ckdb::keyMeta (arrayParent.getKey());
ssize_t metaKeysSizeWithKsGetSize = ckdb::ksGetSize (metaKeys);

// use keyNextMeta()
ssize_t metaKeysSize = 0;
while (ckdb::keyNextMeta (arrayParent.getKey ()))
	metaKeysSize++;
```

If I use the `metaKeysSizeWithKsGetSize` variable, the test `testshell_markdown_yajl` fails, because a 
`"___dirdata:"`-entry gets inserted at the beginning of an array, while this entry is not present when using `keyNextMeta()` for determining the number of metakeys.


```cpp
#ifdef HAVE_LOGGER
   ELEKTRA_LOG_WARNING ("Metakeys size with ksGetSize: %zd, and with keyNextMeta: %zd",
      metaKeysSizeWithKsGetSize, metaKeysSize);
   logKeySet (metaKeys);
#endif
```


I logged the two variants and the content of the KeySet `metaKeys` to a logfile.
This are the interesting results:

```
Metakeys size with ksGetSize: 2, and with keyNextMeta: 1 (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:241)
	“meta:/array”: “#0” (in logKeySet at src/plugins/directoryvalue/log.hpp:38)
	“meta:/binary”: “” (in logKeySet at src/plugins/directoryvalue/log.hpp:38)
```

Another interesting observation is that, if I call the function `logKeySet (metaKeys)`, the test `testshell_markdown_yajl` also fails.
If I comment the line out, the test succeeds.


The next step was to print the metakeys manually with the `ckdb`-functions.

```cpp
#ifdef HAVE_LOGGER

   const ckdb::Key * curMeta;
   elektraCursor it = 0;
   while ((curMeta = ckdb::keyNextMeta (arrayParent.getKey ())))
   {
      ELEKTRA_LOG_WARNING("KEYNEXTMETA\n%zd. Metakey - name: %s, val: %s",
         it, ckdb::keyName(curMeta), ckdb::keyIsBinary(curMeta) ? "binary":
         ckdb::keyString(curMeta));
      metaKeysSize++;
    }

   ELEKTRA_LOG_WARNING("Metakeys size with ksGetSize: %zd, and with keyNextMeta: %zd",
      metaKeysSizeWithKsGetSize, metaKeysSize);

   for (it = 0; it < metaKeysSizeWithKsGetSize; ++it)
   {
      curMeta = ckdb::ksAtCursor(metaKeys, it);
      ELEKTRA_LOG_WARNING("KSATCURSOR\n%zd. Metakey - name: %s, val: %s", it,
         ckdb::keyName(curMeta), ckdb::keyIsBinary(curMeta) ? "binary" :
         ckdb::keyString(curMeta));
   }
#endif
```

With this approach, the test `testshell_markdown_yajl` succeeded.
The logfile had the following content:
```
KEYNEXTMETA@0. Metakey - name: meta:/binary, val:  (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:250)
Metakeys size with ksGetSize: 2, and with keyNextMeta: 1 (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:255)
KSATCURSOR@0. Metakey - name: meta:/array, val: #0 (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:262)
KSATCURSOR@1. Metakey - name: meta:/binary, val:  (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:262)

...

KEYNEXTMETA@0. Metakey - name: meta:/binary, val:  (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:250)
Metakeys size with ksGetSize: 2, and with keyNextMeta: 1 (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:255)
KSATCURSOR@0. Metakey - name: meta:/array, val: #1 (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:262)
KSATCURSOR@1. Metakey - name: meta:/binary, val:  (in splitEmptyArrayParents at src/plugins/directoryvalue/directoryvalue_delegate.cpp:262)
```

So it seems that the internal cursor for the metakeys is not before, but at the first element and so `keyNextMeta ()` returns the 2nd element.
That's especially weird, because in the original code the `parent.rewindMeta ()` function was called before the loop...

Reply:
@flo91 please also document the ways how to iterate metadata, ideally in an tutorial and in the API docu.